plugins/hydra_nevergrad_sweeper/tests/__init__.py

```diff
@@ -1 +1,37 @@
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+from typing import Any, List
+
+from hydra.core.override_parser.types import (
+    ChoiceSweep,
+    Override,
+    OverrideType,
+    RangeSweep,
```

Comment:
Looks like you did not use interval()
This make me wonder how your tests are even working, interval() is critical for the Nevergrad sweeper.

Reply:
according to the [nevergrad doc](https://hydra.cc/docs/next/plugins/nevergrad_sweeper/), [these params](https://github.com/facebookresearch/hydra/blob/master/plugins/hydra_nevergrad_sweeper/example/config.yaml#L16-L37) are all `ranges`, they are all translated into the same Nevergrad `Scalar` object.  I think we should be consistent with Nevergrad in this case.

Reply:
These are he docs for this plugin, not Nevergrad's docs. They will need to be changed. This change is a dramatic change in API for this plugin.

Please pause work on this until we talk. I think we have very different ideas of what this task is about.

Reply:
Cool, I will dig deeper into nevergrad's doc then.

Reply:
The plugin is creating an interface to Nevergrad.
That interface is using the Hydra grammar, and should be consistent with that, not with the underlying implementation.
This is a substantial shift, before each plugin was free to define it's own command line semantics. this is not the case anymore. No both optimizer plugins, Nevergrad and Ax need to use the command line parsing provided by Hydra, and they should do it in a consistent way to allow users to easily switch to the optimizer that works best for them.

The plugins are responsible for translating from the language Hydra is using to express the desire of the user to the language used by the underlying framework (Ax, Nevergrad etc).

Reply:
I see, yea, that makes sense! Thx

Reply:
@omry, I read the nevergrad doc more thoroughly and have created the following mapping, does it look more reasonable now?  
```
# discrete
   choice(1,2,3)  ==>  ng.p.Choice([["Helium", "Nitrogen", "Oxygen"]])
   tag(transition,choice(2,3,4))  ==> ng.p.TransitionChoice(["Solid", "Liquid", "Gas"])
   tag(transition, range(1, 11, 2)) ==> ng.p.TransitionChoice([1, 3, 5 ...])
   range(1,5) ==>  ng.p.Scalar(lower=1, upper=12).set_integer_casting(),
# continuous
   interval(1.0,5.0) ==> np.p.Scalar(lower=1.0, upper=5.0)
   tag(log, interval(0.1, 1)) ==> np.p.Log(lower=0.1, upper=1)
```

Reply:
Looking better.

Looking at the existing tests:
```
        ("blu,blublu", ng.p.Choice, str),
        (["blu", "blublu"], ng.p.Choice, str),
        ("0,1,2", ng.p.TransitionChoice, float),
        ([0, 1, 2], ng.p.TransitionChoice, float),
        ("int:0,1,2", ng.p.TransitionChoice, int),
        ("str:0,1,2", ng.p.Choice, str),
        ("0.0,12.0,2.0", ng.p.Choice, float),
        ("int:1:12", ng.p.Scalar, int),
        ("1:12", ng.p.Scalar, float),
        ("log:0.01:1.0", ng.p.Log, float),
        ("blublu", str, str),
```

It looks like any numeric choice is considered transitional, and string choices are considered categorical (normal choice).
I think you should not introduce a new tag for this. we can keep the same semantics.

Reply:
I think the tests are a bit misleading here:

According to the nevergrad doc here https://facebookresearch.github.io/nevergrad/machinelearning.html?highlight=transitionchoice
`arg1 = ng.p.TransitionChoice(["a", "b"])  # 1st arg. = positional discrete argument`

Strings can also be transitional.

Reply:
What does transitional actually mean? that there is an initial guess in the center of the range?

Reply:
Choice and TransitionChoice behaves differently: 

[Choice](https://facebookresearch.github.io/nevergrad/parametrization_ref.html?highlight=transition#nevergrad.p.Choice) draws values randomely

[TransitionChoice](https://facebookresearch.github.io/nevergrad/parametrization_ref.html?highlight=transition#nevergrad.p.TransitionChoice): Ordered categorical parameter, The chosen parameter is drawn using transitions between current choice and the next/previous ones. Although I'm not sure if it's common to sweep non-numerical  ordered values, but it is something Nevergrad supports.

Reply:
Ax also support choices. I am not sure if they expose control over the sampling method.
@jrapin, does it make sense to you to use a tag to indicate transitional vs a regular choice?

```python
choice(a,b,c)  =>  ng.p.Choice(["a", "b", "c"])
tag(transition, choice(a,b,c)) ==> ng.p.TransitionChoice(["a","b","c"])
```

Reply:
To give full(er) control on the parametrization yes, but is it really useful for the plugin, not so sure.
If you are at the point where you want absolute control, better use nevergrad directly.
So I'd probably stick to `TransitionChoice` for ordered lists, and `Choice` for anything else.
If that becomes necessary, we can always add tag support later on to override the default behavior?

> I think the tests are a bit misleading here

Indeed, I should have added other tests, `[3, 2, 1]` is considered as unordered in the plugin since it is not strictly increasing.

Reply:
> This is what I am thinking after reading the comments:
> 
> ```python
> # regardless of the choice values, choice is always unordered p.Choice:
> choice(a,b,c)  =>  ng.p.Choice(["a", "b", "c"]) 
> 
> # We can support forcing a choice to be ordered by tagging it:
> # (I prefer order over transition(al). at least to me it's clearer.
> tag(ordered, choice(a,b,c)) ==> ng.p.TransitionChoice(["a","b","c"])
> 
> # ranges are always ordered:
> range(1,5)  =>  ng.p.TransitionChoice([1,2,3,4])
> 
> # unless they are shuffled (There is a flag shuffle in the RangeSweep object.
> shuffle(range(1,5)) ->  =>  ng.p.Choice([4,2,1,3]) 
> 
> # intervals are scalars:
> # Note: by intervals are always interpreted as floats (even for int start and end values).
> interval(0,1)     -> RangeSweep(start=0.0, end=1.0) -> ng.p.Scalar(lower=0.0, upper=1.0)
> interval(0.0,1.0) -> RangeSweep(start=0.0, end=1.0) -> ng.p.Scalar(lower=0.0, upper=1.0)
> 
> # a user can cast the interval to int to override that:
> int(interval(0,1) -> RangeSweep(start=0, end=1) -> ng.p.Scalar(lower=0.0, upper=1.0).set_integer_casting()
> ```
> 
> If everyone think this makes sense we can go ahead with it.

lgtm. I will go ahead with the implementation then.


Thx @jrapin  for taking a look! Hope you enjoy your time off 🙃